include/dxc/Support/WinAdapter.h

```diff
@@ -38,21 +44,49 @@
 //                             Begin: Macro Definitions
 //
 //===----------------------------------------------------------------------===//
-#define C_ASSERT(expr) static_assert((expr), "")
+// If MinGW, we need to define the uuid macro but nothing else.
+#ifdef __MINGW32__
+#include <guiddef.h>
+
+// Since type declaration headers (dxcapi.h) are pulled in for multiple compilation units (e.g.
+// both dxcapi.use.cpp and FileIOHelper.cpp pull it in), and because MinGW UUID declarations need
+// to export a customization of the __mingw_uuidof<type>, we would run into duplicate symbols
+// across compilation units. To workaround this, we instead define CROSS_PLATFORM_UUIDOF as nothing
+// and then invoke MINGW_UUIDOF in dxcapi.use.cpp explicitly for each COM interface we wish to
+// export as our public API.
+#define CROSS_PLATFORM_UUIDOF(type, spec)
+
+#define MINGW_UUIDOF(type, spec)                                              \
+    extern "C++" {                                                            \
+    struct __declspec(uuid(spec)) type;                                       \
+    template<> const GUID &__mingw_uuidof<type>() {                           \
+        static constexpr IID __uuid_inst = guid_from_string(spec);            \
+        return __uuid_inst;                                                   \
+    }                                                                         \
+    template<> const GUID &__mingw_uuidof<type*>() {                          \
+        return __mingw_uuidof<type>();                                        \
+    }                                                                         \
+    }
+
```

Comment:
@kusma This is not using constexpr and would case compiling issues when using __uuidof in static_assert